### ISSUE TITLE
URL in request missing leading /

### ISSUE THREAD

**Issue by runejuhl** [babelouest/ulfius#113]
Hi,

I noticed that uwsc 0.9 (from Ubuntu disco repos) apparently splits the URL wrong which means that the leading slash (`/`) in the request is missing. As far as I can tell this this breaks the HTTP spec. See e.g. https://stackoverflow.com/a/27638445

Running uwsc like this:

```sh
uwsc --add-header='x-ssl-client-cn: omg.lol' http://localhost:9090/ws
```

results in a tcpdump snippet like so:

~~~
23:20:54.405380 IP localhost.38000 > localhost.9090: Flags [P.], seq 1:18, ack 1, win 512, options [nop,nop,TS val 4046921048 ecr 4046921048], length 17
E..E.}@.@..3.........p#.9O|...^c.....9.....
.7.X.7.XGET ws HTTP/1.1
~~~

I haven't dug into the source of uwsc so I'm not sure if it's been patched already, but I couldn't find any similar issues.

A workaround is to just prepend the path with an extra slash, like so:

```sh
uwsc --add-header='x-ssl-client-cn: omg.lol' http://localhost:9090//ws
```

Version information:
~~~
$ apt policy uwsc
uwsc:
  Installed: 2.5.2-4
  Candidate: 2.5.2-4
  Version table:
 *** 2.5.2-4 500
        500 http://dk.archive.ubuntu.com/ubuntu disco/universe amd64 Packages
        100 /var/lib/dpkg/status
~~~

**Comment by babelouest** [comment]
Hello @runejuhl , thanks for the bug report!

Indeed ulfius websocket client library has a bug [here](https://github.com/babelouest/ulfius/blob/master/src/u_websocket.c#L567).
I fixed it in the branch [2.5](https://github.com/babelouest/ulfius/blob/2.5/src/u_websocket.c#L567) as well as the next branch 2.6 with the patch.

Can you test the branch [2.5](https://github.com/babelouest/ulfius/tree/2.5) and give me some feedback if it's properly fixed?

**Comment by runejuhl** [comment]
Whoa, you're fast :smile: 

Hm, just tried compiling it from scratch:

```sh
mkdir build && cd build
cmake -DWITH_WEBSOCKET=on .. 
make
./uwsc --add-header='x-ssl-client-cn: omg.lol' ws://localhost:9091/ws
```

Running that gives the same result as before with the leading slash missing:

~~~
$ socat stdio tcp-listen:9091,fork,reuseaddr
GET ws HTTP/1.1
Host: localhost
Upgrade: websocket
Connection: Upgrade
Origin: ws://localhost
x-ssl-client-cn:  omg.lol
Sec-WebSocket-Version: 13
Upgrade: websocket
Connection: Upgrade
Content-Length: 0
User-Agent: Ulfius Websocket Client Framework/2.5.2
Sec-WebSocket-Key: QW1hUHlnT2pxdERZMUhIUg==
~~~

**Comment by babelouest** [comment]
This shouldn't be...
Did you checkout the branch 2.5?

```shell
$ git clone https://github.com/babelouest/ulfius.git
$ git checkout 2.5
$ mkdir build && cd build
$ cmake -DWITH_WEBSOCKET=on .. && make
```

**Comment by runejuhl** [comment]
Yes, that's the tag I've checked out:

~~~
$ git log -n1
commit dc209dd255a54492142481514178e63ea639bd82 (HEAD -> 2.5, github/2.5)
Author: Nicolas Mora <github@babelouest.org>
Date:   Thu May 9 20:45:45 2019 -0400

    Fix missing leading / in websocket client url, #113
~~~

**Comment by runejuhl** [comment]
Just tried again with a clean build folder, same result.

**Comment by babelouest** [comment]
Something seems weird...
Maybe you still use ulfius library provided by Ubuntu?

If using the following command doesn't change anything:

```
$ LD_LIBRARY_PATH=. ./uwsc --add-header='x-ssl-client-cn: omg.lol' ws://localhost:9091/ws # under ulfius/build
```

Can you run the following commands and give me the result?

```shell
$ cd tools/uwsc
$ make debug
$ LD_LIBRARY_PATH=../../src/ ./uwsc --add-header='x-ssl-client-cn: omg.lol' ws://localhost:9091/ws
```

**Comment by runejuhl** [comment]
You're right, sorry for being a bit dense here :)

```sh
LD_LIBRARY_PATH=. ./uwsc --add-header='x-ssl-client-cn: omg.lol' ws://localhost:9091/ws
```

~~~
GET /ws HTTP/1.1
Host: localhost
Upgrade: websocket
Connection: Upgrade
Origin: ws://localhost
x-ssl-client-cn:  omg.lol
Sec-WebSocket-Version: 13
Upgrade: websocket
Connection: Upgrade
Content-Length: 0
User-Agent: Ulfius Websocket Client Framework/2.5.4
Sec-WebSocket-Key: NWFUR3B2cWJZcmwwdFBFUA==
~~~

**Comment by babelouest** [comment]
Glad to know :-)

Thanks for the bug report!

**Comment by babelouest** [comment]
I'm closing this issue, the fix will be present in the next 2.5 release.
